cmd/snap-confine/snap-confine.apparmor.in

```diff
@@ -404,4 +382,116 @@
+    # distribution to distribution. The variants here represent different
+    # locations of snap mount directory across distributions.
+    /{,var/lib/snapd/}snap/core/*/usr/lib/snapd/snap-update-ns Cxr -> snap_update_ns,
+
+    # ...snap-confine is, conceptually, re-executing and uses snap-update-ns
+    # from the core snap but we are already inside the constructed mount
+    # namespace. Here the apparmor kernel module re-constructs the path to
+    # snap-update-ns using the "hostfs" mount entry rather than the more
+    # "natural" /snap mount entry but we have no control over that.  This is
+    # reported as (LP: #1716339). The variants here represent different
+    # locations of snap mount directory across distributions.
+    /var/lib/snapd/hostfs/{,var/lib/snapd/}snap/core/*/usr/lib/snapd/snap-update-ns Cxr -> snap_update_ns,
+
+    profile snap_update_ns (attach_disconnected) {
+        # The next four rules mirror those above. We want to be able to read
+        # and map snap-update-ns into memory but it may come from a variety of places.
+        /usr/lib{,exec}/snapd/snap-update-ns mr,
+        /var/lib/snapd/hostfs/usr/lib{,exec}/snapd/snap-update-ns mr,
+        /{,var/lib/snapd/}snap/core/*/usr/lib/snapd/snap-update-ns mr,
+        /var/lib/snapd/hostfs/{,var/lib/snapd/}snap/core/*/usr/lib/snapd/snap-update-ns mr,
+
+        # Allow reading the dynamic linker cache.
+        /etc/ld.so.cache r,
+        # Allow reading, mapping and executing the dynamic linker.
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}ld-*.so mrix,
+        # Allow reading and mapping various parts of the standard library and
+        # dynamically loaded nss modules and what not.
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libc{,-[0-9]*}.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libdl{,-[0-9]*}.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libgcc_s.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libpcre.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libpthread{,-[0-9]*}.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libresolv{,-[0-9]*}.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}librt{,-[0-9]*}.so* mr,
+        /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libselinux.so* mr,
+
+        # Allow reading the command line (snap-update-ns uses it in pre-Go bootstrap code).
+        @{PROC}/@{pid}/cmdline r,
+
+        # Allow reading the os-release file (possibly a symlink to /usr/lib).
+        /{etc/,usr/lib/}os-release r,
+
+        # Allow creating/grabbing various snapd lock files.
+        /run/snapd/lock/*.lock wrk,
+
+        # Allow reading stored mount namespaces,
+        /run/snapd/ns/ r,
+        /run/snapd/ns/*.mnt r,
+
+        # Allow reading per-snap desired mount profiles. Those are written by
+        # snapd and represent the desired layout and content connections.
+        /var/lib/snapd/mount/snap.*.fstab r,
```

Comment:
In the snap-confine profile, this rule was `/{tmp/snap.rootfs_*/,}var/lib/snapd/mount/*.fstab r,`. To questions:

1. why didn't you use this rule for snap-update-ns?
2. why didn't you remove this rule from snap-confine?

Reply:
I also noticed this rule was in snap-confine's profile, that seems like it can be removed:
```
    /run/snapd/ns/*.fstab rw,
```

Reply:
Both of the rules from snap-confine can be removed now. I will make that happen.

Reply:
In `snap-update-ns` we only deal with paths after pivot_root so the rule can be simpler.

Reply:
I removed the rules from snap-confine now.